src/order/antisymmetrization.lean

```diff
@@ -0,0 +1,175 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import order.hom.basic
+
+/-!
+# Turning a preorder into a partial order
+
+This file allows to make a preorder into a partial order by quotienting out the elements `a`, `b`
+such that `a ≤ b` and `b ≤ a`.
+
+`antisymmetrization` is a functor from `Preorder` to `PartialOrder`. See `Preorder_to_PartialOrder`.
+
+## Main declarations
+
+* `le_equiv`: The antisymmetrization relation. `le_equiv a b` means that `a` and `b` are less than
+  each other.
+* `antisymmetrization α`: The quotient of `α` by `le_equiv`. Even when `α` is just a preorder,
+  `antisymmetrization α` is a partial order.
+
+## TODO
+
+The quotient boilerplate is here because the quotient API requires a `setoid` instance while
+`le_equiv.setoid` definitely can't be a global instance. Quotients should work better in Lean 4 with
+that regard.
+-/
+
+open order_dual
+
+variables {α β : Type*}
+
+section has_le
+variables [has_le α]
+
+/-- The antisymmetrization relation. -/
+def le_equiv (a b : α) : Prop := a ≤ b ∧ b ≤ a
+
+lemma le_equiv.symm {a b : α} : le_equiv a b → le_equiv b a := and.symm
+
+lemma le_equiv_refl [is_refl α (≤)] (a : α) : le_equiv a a := ⟨refl _, refl _⟩
+
+lemma le_equiv.trans [is_trans α (≤)] {a b c : α} (hab : le_equiv a b) (hbc : le_equiv b c) :
+  le_equiv a c :=
+⟨trans hab.1 hbc.1, trans hbc.2 hab.2⟩
+
+instance le_equiv.decidable_rel [@decidable_rel α (≤)] : @decidable_rel α le_equiv :=
+λ _ _, and.decidable
+
+end has_le
+
+section preorder
+variables [preorder α] [preorder β] {a b : α}
+
+lemma le_equiv.image {f : α → β} (hf : monotone f) {a b : α} (h : le_equiv a b) :
+  le_equiv (f a) (f b) :=
+⟨hf h.1, hf h.2⟩
+
+variables (α)
+
+/-- The antisymmetrization relation as an equivalence relation. -/
+@[simps, reducible] def le_equiv.setoid : setoid α :=
+⟨le_equiv, le_equiv_refl, λ _ _, le_equiv.symm, λ _ _ _, le_equiv.trans⟩
+
+local attribute [instance] le_equiv.setoid
```

Comment:
I don't like this `local attribute [instance] *setoid` design: this way you can't use two constructions that rely on this for the same type. We have quite a few definitions/lemmas that take `{s : setoid α}` instead of `[setoid α]` and I think that we should use this API in case of non-canonical setoids.

Reply:
This requires some modification to the `quotient` API. Tell me if you think this should be split off this PR.

Reply:
Iirc, Lean 4 apparently also doesn't have `setoid` as a class. I think it would be good to split such a refactor off from this PR.

Reply:
#11824